File: finding-seams/csharp/FindingSeams/Reminders/ReminderSender.cs
using System.Globalization;

namespace FindingSeams.Reminders;

public class ReminderSender
{
    private readonly int _retryCount = 3;

    public async Task SendAllAsync(IReadOnlyList<Reminder> reminders)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        foreach (var reminder in reminders)
        {
            var sendAt = ToTimestamp(reminder.SendAt);
            if (sendAt <= now)
                await RetryReminderSendingAsync(reminder).ConfigureAwait(false);
        }
    }

    private async Task RetryReminderSendingAsync(Reminder reminder)
    {
        var currentCount = 0;
        while (true)
        {
            try
            {
                await SendsReminders.Instance.SendAsync(reminder).ConfigureAwait(false);
                return;
            }
            catch
            {
                Console.WriteLine("Error sending reminder:  Retrying...");
            }

            var cond = currentCount < _retryCount;
            currentCount++;
            if (!cond)
                break;
        }
    }

    private static long ToTimestamp(object? value) =>
        value switch
        {
            null => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            DateTimeOffset dto => dto.ToUnixTimeMilliseconds(),
            DateTime dt => new DateTimeOffset(dt).ToUnixTimeMilliseconds(),
            long l => l,
            int i => i,
            string s when s.All(char.IsDigit) => long.Parse(s, CultureInfo.InvariantCulture),
            string s when DateTimeOffset.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind,
                out var p) => p.ToUnixTimeMilliseconds(),
            _ => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
        };
}

File: finding-seams/kotlin/src/main/kotlin/findingseams/remindersender/ReminderSender.kt
package findingseams.remindersender

import java.time.Instant
import java.time.format.DateTimeFormatter

class ReminderSender(
    private val retryCount: Int = 3,
) {
    suspend fun sendAll(reminders: List<Reminder>) {
        val now = System.currentTimeMillis()
        for (reminder in reminders) {
            val sendAtTime = toTimestamp(reminder.sendAt)
            if (sendAtTime <= now) {
                retryReminderSending(reminder)
            }
        }
    }

    private suspend fun retryReminderSending(reminder: Reminder) {
        var currentCount = 0
        while (true) {
            try {
                SendReminders.send(reminder)
                return
            } catch (_: Exception) {
                println("Error sending reminder:  Retrying...")
            }
            val cond = currentCount < retryCount
            currentCount++
            if (!cond) break
        }
    }

    private fun toTimestamp(value: Any?): Long {
        when (value) {
            null -> return System.currentTimeMillis()
            is Instant -> return value.toEpochMilli()
            is Number -> return value.toLong()
            is String -> {
                if (value.all { it.isDigit() }) {
                    return value.toLong()
                }
                val parsed = runCatching { Instant.parse(value) }.getOrNull()
                    ?: runCatching {
                        java.time.LocalDateTime.parse(value, DateTimeFormatter.ISO_DATE_TIME).atZone(
                            java.time.ZoneId.systemDefault(),
                        ).toInstant()
                    }.getOrNull()
                return parsed?.toEpochMilli() ?: System.currentTimeMillis()
            }
            else -> return System.currentTimeMillis()
        }
    }
}
